### PR TITLE
Port LicenseHeadersTask to java and remove ant layer

### PR DESCRIPTION
**Pull request by breskeby**
- This ports Licenseheaders task from using groovies antbuilder to rely on the rat ant task to use 
rat directly which allows us to port this completely to java. We also refactored rat to generate xml instead of plain report file. This is 
- Adding basic integTest coverage for `licenseheaders` plugin
a) more efficient as plain file generation is based on running an xslt transformation on top of the xml output (via ant rat in a different thread which results in faster task execution as seen in the attached profiler report 
b) easier machine readable

By using rat directly we can avoid parsing the log file for reporting. 

- Ported more buildSrc sources from groovy to java which has been blocked from porting by LicenseHeadersTask

Screenshot 2020-11-23 at 17 55 32

- red: `./gradlew licenseHeaders` from es root folder (clean repo) before refactoring
- green: `./gradlew licenseHeaders` from es root folder (clean repo) after refactoring

### CODE REVIEW

File: buildSrc/build.gradle [hunk 1]
```diff
@@ -21,7 +21,7 @@ import org.gradle.internal.jvm.Jvm
 import org.gradle.util.GradleVersion
 
 plugins {
-  id 'java-gradle-plugin'
+  id 'groovy-gradle-plugin'
```
Comment (mark-vieira) [inline buildSrc/build.gradle:24]: What's the difference here? Do we still need to expliclty apply the `groovy` plugin?

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/InternalPrecommitTasks.java.new [hunk 2]
```diff
@@ -0,0 +1,66 @@
+package org.elasticsearch.gradle.internal.precommit;
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/InternalPrecommitTasks.java.new:1]: Looks like this snuck in by accident.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/StandaloneTestPlugin.java.new [hunk 3]
```diff
@@ -0,0 +1,48 @@
+package org.elasticsearch.gradle.test;
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/StandaloneTestPlugin.java.new:1]: This also looks like it should be removed.

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/LicenseHeadersTask.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+     * source file is compatible with the codebase but we do not want to add the license to the list of approved headers (to avoid the
+     * possibility of inadvertently using the license on our own source files).
+     */
+    @Input
+    private List<String> excludes = new ArrayList<String>();
+    /**
+     * Additional license families that may be found. The key is the license category name (5 characters),
+     * followed by the family name and the value list of patterns to search for.
+     */
+    protected Map<String, String> additionalLicenses = new HashMap<String, String>();
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/LicenseHeadersTask.java:238]: Shouldn't this be tracked as an input as well?

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/LicenseHeadersTask.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Checks files for license headers.
+ */
+public abstract class LicenseHeadersTask extends DefaultTask {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/LicenseHeadersTask.java:66]: Anything keeping us from making this cacheable?

**Review by mark-vieira — APPROVED** [review]
Few comments but otherwise LGTM. That's a great performance bump we get from not delegating to Ant anymore. Nice work.

File: buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/LicenseHeadersTask.java (hunk 5 above)
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/internal/precommit/LicenseHeadersTask.java:66]: No I don't think so. will do that

File: buildSrc/build.gradle (hunk 1 above)
Comment (breskeby) [reply buildSrc/build.gradle:24]: A leftover from figuring out why the test kit integ tests didnt work. should be revertible. will check

**Comment by breskeby** [comment]
@elasticmachine test this please

**Pull request merged**